test: interrupt: add some nop operations to make sure the interrupt is handled.

chen-png opened the pull request:
On X86 platforms, the interrupt trigger method has been changed
from using INT command to using APIC IPI, we need to make sure
the IPI interrupt is handled before do our check, so add some
nop operations.

Also change the test interrupt line number to a bigger one, because
the low number usually will be used by other devices.

Commit:
test: interrupt: change the test interrupt line to a bigger one.

change the test interrupt line number to a bigger one, because
the low number usually will be used by other devices.

Signed-off-by: Chen Peng1 <peng1.chen@intel.com>

andyross reviewed (changes requested):
This looks wrong to me, or at least needs more explanation.  The intent of the trigger_irq() function is clearly that it should be synchronous.  If it's not, then you want to add your needed delay code there.

But on most platforms, that trigger is indeed synchronous!  Almost everywhere, it works by flagging an interrupt latch register that then interrupts the code that did it (though I guess potentially with some cycles of delay if it's an external controller and not e.g. an x86 INT instruction).  You really shouldn't need 5us of delay to make this work, and if you do that implies something else is going wrong.

Which platform needs this?  I think we need to do more debugging here to figure out what the root cause is.

Finally: note that k_busy_wait() on some architectures involves masking interrupts (I forget the details, but I remember a bug to that effect once).  So this might not work at all everywhere.

chen-png commented:
> flagging an interrupt latch register that then interrupts the code that did it (though I guess potentially with some cycles of delay if it's an external controller and not e.g. an x86 INT instruction)

this is on an internal x86 platform, and the interrupt trigger method on x86 has been changed from INT command to using LOAPIC IPI.
if it used INT command to trigger interrupt,  it worked well without delay, but if using IPI, it failed without this delay.

enjiamai reviewed (changes requested):
> Which platform needs this? I think we need to do more debugging here to figure out what the root cause is.
> 

Totally agree with you. @andyross , we have tried to add " __asm__ volatile ("wbinvd");"  to invalidate the cache before read the memory and did not see the error anymore. Could you please give us some further suggestions about how to fix this?

andyross commented:
>  if it used INT command to trigger interrupt, it worked well without delay, but if using IPI, it failed without this delay.

Then can we add the delay in a platform-specific way?  Again, the clear intent of that test API is that it should be synchronous, returning after the interrupt was processed.

And if it's just x86, then k_busy_wait() is fine to use, as it just spins on the timer driver, and shouldn't affect interrupt delivery (no locking is needed for HPET or APIC_TSC drivers).

chen-png commented:
> the clear intent of that test API is that it should be synchronous

yes, the test logic should be synchronous, but I am not sure it is now.

@andyross **Does it need some cycles to wait for the interrupt handled if using IPI?**
I pasted some assembly code below, after changing to use IPI in trigger_irq, it will write specific test value into ICR register to trigger interrupt and  then invoke "ret" to return to the main thread, at this point, does it make sure the interrupt was handled before we check global variable "handler_has_run" in main thread?
```
interrupt handler:
ff20063d <trigger_irq>:
        ipi |= vector;
ff20063d:       0f b6 c0                movzbl %al,%eax
ff200640:       80 cc 40                or     $0x40,%ah
        __asm__ volatile("movl %1, %0"
ff200643:       8b 15 00 03 e0 fe       mov    0xfee00300,%edx
        while (x86_read_xapic(LOAPIC_ICRLO) & LOAPIC_ICR_BUSY) {
ff200649:       81 e2 00 10 00 00       and    $0x1000,%edx
ff20064f:       75 f2                   jne    ff200643 <trigger_irq+0x6>
        __asm__ volatile("movl %0, %1"
ff200651:       89 15 10 03 e0 fe       mov    %edx,0xfee00310
ff200657:       a3 00 03 e0 fe          mov    %eax,0xfee00300
ff20065c:       c3                      ret

main thread:
        trigger_irq(vector_num);
ff2006e1:       89 d8                   mov    %ebx,%eax
ff2006e3:       e8 55 ff ff ff          call   ff20063d <trigger_irq>

        zassert_equal(handler_has_run, 1,
ff2006e8:       a1 4c 99 20 ff          mov    0xff20994c,%eax
ff2006ed:       50                      push   %eax
ff2006ee:       68 98 7f 20 ff          push   $0xff207f98
ff2006f3:       68 e0 79 20 ff          push   $0xff2079e0
ff2006f8:       6a 6c                   push   $0x6c
ff2006fa:       68 dd 7e 20 ff          push   $0xff207edd
ff2006ff:       68 c8 7f 20 ff          push   $0xff207fc8
```

andyross commented:
On x86 at least, the trigger_irq() interrupt was always synchronous because the INT instruction is synchronous (no instructions execute before the interrupt is entered).  On other platforms, I assume it is, but we'd have to check.  So you can't change to an async mechanism without breaking compatibility and forcing the workaround here.

The right way to do this is to write the code such that you are guaranteed the interrupt has fired before trigger_irq() returns in thread context.  It might be enough to just put in a bunch of NOP instructions or a spin loop, or maybe it would be better to have the ISR set a flag for which you could poll, etc...

Commit:
tests: interrupt: add some nop operations in trigger_irq function.

On X86 platforms, the interrupt trigger method has been changed
from using INT command to using APIC IPI, we need to make sure
the IPI interrupt is handled before do our check, so add some
nop operations.

Signed-off-by: Chen Peng1 <peng1.chen@intel.com>

Commit:
tests: interrupt: remove unused macro TRIGGER_IRQ_INT.

On X86 platforms, the interrupt trigger method has been
changed to use APIC IPI, we don't use INT command to trigger
interrupt, so remove this unused macro.

Signed-off-by: Chen Peng1 <peng1.chen@intel.com>

chen-png commented:
@andyross @dcpleung @peter-mitsis @enjiamai Could you help review this? thanks.

on x86, now we don't use INT instruction to trigger interrupt, we use APIC IPI, this is not synchronous, it needs some cycles.

cfriedt commented:
@chen-png, @andyross - would this also apply to v2.7?

andyross commented:
Dunno, I guess my sense is that with changes to test code we should only backport if there are known failures with someone actually running those tests.  Not sure that's true of x86 hardware with 2.7?

chen-png commented:
> would this also apply to v2.7?

No

> only backport if there are known failures with someone actually running those tests

+1, so v3.1-branch is enough
